File: python/src/material.py
import numpy as np

class Material:
    def __init__(self,id,style,param):
        self.id = id
        self.style = style

        self.set_param(param)

    def print(self):
        print(self.id,":",self.style,self.param)

    def set_param(self,param):
        self.param = param

        if self.style == "vs_vp_rho":
            vs,vp,rho = param
            self.rmu = rho*vs*vs
            self.rlambda = rho*vp*vp - 2.*self.rmu
            self.rho = rho

        elif self.style == "nu_vp_rho":
            nu,vp,rho = param
            self.rmu = rho/2*vp**2*(1-2*nu)/(1-nu)
            self.rlambda = rho*nu*vp**2/(1-nu)
            self.rho = rho

        elif self.style == "nu_vs_rho":
            nu,vs,rho = param
            self.rmu = rho*vs**2
            self.rlambda = 2*nu/(1-2*nu) * self.rmu
            self.rho = rho

        elif self.style == "nu_E_rho":
            nu,E,rho = param
            self.rmu = E/2.0/(1+nu)
            self.rlambda = 2*nu/(1-2*nu) * self.rmu
            self.rho = rho

        elif self.style == "spring_normal":
            self.rho = 0.0
            self.kv,self.kh,n0,n1 = param

            norm = np.sqrt(n0*n0+n1*n1)
            n0,n1 = n0/norm,n1/norm

            self.R = np.zeros([2,2], dtype=np.float64)
            self.R[0,0],self.R[0,1] =  n0, n1
            self.R[1,0],self.R[1,1] = -n1, n0

        elif self.style == "slider_normal":
            self.rho = 0.0
            n0,n1 = param

            norm = np.sqrt(n0*n0+n1*n1)
            n0,n1 = n0/norm,n1/norm

            self.R = np.zeros([2,2], dtype=np.float64)
            self.R[0,0],self.R[0,1] =  n0, n1
            self.R[1,0],self.R[1,1] = -n1, n0

        elif self.style == "ep_Li":
            self.rho = param[0]
            self.param = param[1:]

        elif self.style == "ep_eff_Li":
            self.rho = param[0]
            self.param = param[1:]

            self.rho_w = 1000.0
            self.Kw = 2.25e9

            e0 = param[4]
            n0 = e0/(1 + e0)
            self.rho = (1-n0)*param[0] + n0*self.rho_w
            # print("rho_t:",self.rho)

        elif self.style == "ep_GHES":
            self.rho = param[0]
            self.param = param[1:]

    # ---------------------------------------------------------
    def mk_d(self,dof):
        if dof == 1:
            D = np.zeros([2,2],dtype=np.float64)
            D[0,0] = self.rmu
            D[1,1] = self.rmu

        elif dof == 2:
            D = np.zeros([3,3],dtype=np.float64)
            D[0,0],D[0,1] = self.rlambda + 2*self.rmu, self.rlambda
            D[1,0],D[1,1] = self.rlambda, self.rlambda + 2*self.rmu
            D[2,2] = self.rmu

        elif dof == 3:
            D = np.zeros([5,5],dtype=np.float64)
            D[0,0],D[0,1] = self.rlambda + 2*self.rmu, self.rlambda
            D[1,0],D[1,1] = self.rlambda, self.rlambda + 2*self.rmu
            D[2,2] = self.rmu
            D[3,3] = self.rmu
            D[4,4] = self.rmu

        return D

    def mk_d_spring(self):
        D = np.zeros([4,4],dtype=np.float64)
        D[0,0],D[2,0] =  self.kv, -self.kv
        D[0,2],D[2,2] = -self.kv,  self.kv
        D[1,1],D[3,1] =  self.kh, -self.kh
        D[1,3],D[3,3] = -self.kh,  self.kh

        return D

    # ---------------------------------------------------------
    def mk_visco(self,dof):
        mu = 0.001 # [Pa s]

        if dof == 1:
            D = np.zeros([2,2],dtype=np.float64)
            D[0,0] = mu
            D[1,1] = mu

        elif dof == 2:
            D = np.zeros([3,3],dtype=np.float64)
            D[2,2] = mu

        elif dof == 3:
            D = np.zeros([5,5],dtype=np.float64)
            D[2,2] = mu
            D[3,3] = mu
            D[4,4] = mu

        return D

    # ---------------------------------------------------------
    def mk_imp(self,dof):
        vs = np.sqrt(self.rmu/self.rho)
        vp = np.sqrt((self.rlambda +2*self.rmu)/self.rho)

        if dof == 1:
            imp = np.diag([self.rho*vs])
        elif dof == 2:
            imp = np.diag([self.rho*vp,self.rho*vs])
        elif dof == 3:
            imp = np.diag([self.rho*vp,self.rho*vs,self.rho*vs])

        return imp

File: cpp/src/material.cpp
#include "all.h"
#include <Eigen/Core>
#include "material.h"

using EV = Eigen::VectorXd ;
using EM = Eigen::MatrixXd ;

Material::Material () {}
Material::Material (size_t id, std::string style, std::vector<double> param) {
  this->id = id;
  this->style = style;
  this->param = param;

  this->set_param();
}

void Material::set_init(size_t id, std::string style, std::vector<double> param) {
  this->id = id;
  this->style = style;
  this->param = param;

  this->set_param();
}

void Material::print() {
  std::cout << this->id << ": ";
  std::cout << this->style << ", ";
  for (size_t i = 0 ; i < this->param.size() ; i++) {
    std::cout << this->param.at(i) << " ";
  }
  std::cout << "\n";
}

void Material::set_param() {
  if (this->style == "vs_vp_rho") {
    double vs = this->param.at(0);
    double vp = this->param.at(1);
    double rho = this->param.at(2);

    this->rmu = rho*vs*vs;
    this->rlambda = rho*vp*vp - 2.0*this->rmu;
    this->rho = rho;

  } else if (this->style == "nu_vp_rho") {
    double nu = this->param.at(0);
    double vp = this->param.at(1);
    double rho = this->param.at(2);

    this->rmu = rho/2.0*vp*vp*(1.0-2.0*nu)/(1.0-nu);
    this->rlambda = rho*nu*vp*vp/(1.0-nu);
    this->rho = rho;

  } else if (this->style == "nu_vs_rho") {
    double nu = this->param.at(0);
    double vs = this->param.at(1);
    double rho = this->param.at(2);

    this->rmu = rho*vs*vs;
    this->rlambda = 2.0*nu/(1.0-2.0*nu) * this->rmu;
    this->rho = rho;

  } else if (this->style == "nu_E_rho") {
    double nu = this->param.at(0);
    double E = this->param.at(1);
    double rho = this->param.at(2);

    this->rmu = E/2.0/(1.0+nu);
    this->rlambda = 2.0*nu/(1.0-2.0*nu) * this->rmu;
    this->rho = rho;

  } else if (this->style == "spring_normal") {
    this->kv = this->param.at(0);
    this->kh = this->param.at(1);
    double n0 = this->param.at(2);
    double n1 = this->param.at(3);

    double norm = sqrt(n0*n0+n1*n1);
    n0 = n0/norm; n1 = n1/norm;

    this->R = EM::Zero(2,2);
    this->R(0,0) =  n0; this->R(0,1) = n1;
    this->R(1,0) = -n1; this->R(1,1) = n0;
    this->rho = 0.0;

  } else if (this->style == "ep_Li") {
    this->rho = param.at(0);
    for (size_t i=1; i < this->param.size(); i++) {
      this->param_ep.push_back(this->param.at(i));
    }
    
  } else if (this->style == "ep_eff_Li") {
    double e0 = param.at(4);
    double n0 = e0 / (1 + e0);
    this->rho = (1 - n0)*param.at(0) + n0*this->rho_w;
    for (size_t i=1; i < this->param.size(); i++) {
      if (i == 0) {
        this->param_ep.push_back(this->rho);
      } else {
        this->param_ep.push_back(this->param.at(i));
      }
    }
    
  }
}

// ------------------------------------------------------------------- //
EM Material::mk_d(const size_t dof) {
    EM D;

    if (dof == 1) {
      D = EM::Zero(2,2);
      D(0,0) = this->rmu;
      D(1,1) = this->rmu;

    } else if (dof == 2) {
      D = EM::Zero(3,3);
      D(0,0) = this->rlambda + 2.0*this->rmu;
      D(0,1) = this->rlambda;
      D(1,0) = this->rlambda;
      D(1,1) = this->rlambda + 2.0*this->rmu;
      D(2,2) = this->rmu;

    } else if (dof == 2) {
      D = EM::Zero(5,5);
      D(0,0) = this->rlambda + 2.0*this->rmu;
      D(0,1) = this->rlambda;
      D(1,0) = this->rlambda;
      D(1,1) = this->rlambda + 2.0*this->rmu;
      D(2,2) = this->rmu;
      D(3,3) = this->rmu;
      D(4,4) = this->rmu;

    }

    return D;
  }

// ------------------------------------------------------------------- //
EM Material::mk_d_spring() {
    EM D;

    D = EM::Zero(4,4);
    D(0,0) =  this->kv; D(2,0) = -this->kv;
    D(0,2) = -this->kv; D(2,2) =  this->kv;
    D(1,1) =  this->kh; D(3,1) = -this->kh;
    D(1,3) = -this->kh; D(3,3) =  this->kh;

    return D;
  }

// ------------------------------------------------------------------- //
EM Material::mk_visco(const size_t dof) {
    EM D;
    double mu = 0.001; // [Pa s]

    if (dof == 1) {
      D = EM::Zero(2,2);
      D(0,0) = mu;
      D(1,1) = mu;

    } else if (dof == 2) {
      D = EM::Zero(3,3);
      D(2,2) = mu;

    } else if (dof == 2) {
      D = EM::Zero(5,5);
      D(2,2) = mu;
      D(3,3) = mu;
      D(4,4) = mu;

    }

    return D;
  }

// ------------------------------------------------------------------- //
EM Material::mk_imp(const size_t dof) {
    EM imp;
    double vs = sqrt(this->rmu/this->rho);
    double vp = sqrt((this->rlambda + 2.0*this->rmu)/this->rho);

    if (dof == 1) {
      imp = EM::Zero(1,1);
      imp(0,0) = this->rho * vs;

    } else if (dof == 2) {
      imp = EM::Zero(2,2);
      imp(0,0) = this->rho * vp;
      imp(1,1) = this->rho * vs;

    } else if (dof == 3) {
      imp = EM::Zero(3,3);
      imp(0,0) = this->rho * vp;
      imp(1,1) = this->rho * vs;
      imp(2,2) = this->rho * vs;
    }

    return imp;
  }
